### ISSUE TITLE
Best approach for uploading large files?

### ISSUE THREAD

**Issue by escamoteur** [781flyingdutchman/background_downloader#403]
Thanks for your really extensive reply to my previous question. I create a new issue so other's might find this easy.

Currently we use UpChunk to upload large video files to Mux which allows resuming uploads by providing a start-end of the current chunk that I upload. Before every 5MB chunk a call back is called that allows me the next attempted chunk, so that I'm able to resume the upload in case it gets interrupted (currently only uploading while the App is in the foreground.

While studying the API of your package I realized that your progress events only report a percentage but not the actual number of bytes that have been successfully uploaded, so it won't be passive to resume an upload if the app got killed during a background upload.

What is the best approach to upload large files? Just upload them in one big piece and hope that the app won't get killed so the OS will continue uploading it? 
Splitting the file into smaller chunks and enqueue them all with matching range headers? 

or would it be possible to add the actual number of uploaded bytes to the TaskProgressUpdates?

btw. I know from my packages how they can grow :-)

**Comment by 781flyingdutchman** [comment]
There currently isn't a proper RFC standard for resumable and chunked uploads, but there are multiple implementations of mechanisms like you describe. Probably the most common one (and as I understand likely to become an RFC standard) is that used by [Amazon](https://aws.amazon.com/blogs/compute/uploading-large-objects-to-amazon-s3-using-multipart-upload-and-transfer-acceleration/) and [Google](https://cloud.google.com/storage/docs/multipart-uploads) for large file uploads to their respective clouds. 

I have not implemented any of these, given the lack of standardization and the complexity of the associated auth mechanisms, but have recently added the option to add a `Range` header to an upload that will cause only that range to be uploaded (the `Range` header itself is not passed on to the server - but since it is not the same as `Content-Range` you can still pass that like any other header). You can use this to create a chunked upload (and therefore resumption) mechanism, including something similar to what you already use, but until there is a real standard I a don't intend to implement this in the package, especially because this would really need to be implemented on the native side, given this typically is only used for very long running tasks that therefore likely push the app to the suspended stats, so to provide proper progress updates in notifications etc you need to process all that on the native side.

> I realized that your progress events only report a percentage but not the actual number of bytes that have been successfully uploaded

It does include `expectedFileSize` so multiplying that with `progress` gets you that value

> What is the best approach to upload large files? 

I suggest chunking using the `Range` header and using a mechanism like that used by AWS and GCP (or simpler) to identify chunks and stitch them together on the server side. You'll have to keep track of progress across all chunks to see progress for the file as a whole.  The sequential "send a chunk, wait for completion, then send the next" approach isn't well suited for background processing, and doesn't parallelize the upload in the way AWS and GCP allow, which can result in faster processing.

**Comment by 781flyingdutchman** [comment]
One other consideration, discussed in #321 is that if you need authentication, the auth token you used when you enqueued a chunk upload may have expired by the time that chunk actually starts uploading, resulting in a failed upload. Per that discussion, there is currently no mechanism to fix this properly (again, waking up the app upon completion might help, or waking up before starting, but again it would be better to deal with this natively).

**Comment by 781flyingdutchman** [comment]
Final comment: because of the way background uploads are implemented in iOS (by the OS), on iOS only, we actually copy the range of a partial upload to a new temp file and upload that, then delete the temp file. As a result, using partial uploads on iOS will require incremental storage space.  Because this copying happens at enqueue (not at execution), if you enqueue all chunks at once on iOS, you double the storage space for the file to upload.  I know, not great.

**Comment by escamoteur** [comment]
I m currently checking if our video service does offer a call back into our backend on successful uploading. Then we wouldn't need an additional api call after the upload succeeded.

I saw the percentage and the expected file size but isn't this unreliable due to rounding errors? An absolute byte count of already transmitted data would be better. The percentage can be easily calculated.
Am 25. Okt. 2024, 19:10 +0100 schrieb 781flyingdutchman ***@***.***>:

**Comment by 781flyingdutchman** [comment]
> I saw the percentage and the expected file size but isn't this unreliable due to rounding errors?

Possibly, but you'd only be using it for progress indication anyway, right? It'll be plenty accurate for that. You should only use status updates like `TaskStatus.complete` or `progress == 1.0` to determine whether you are done with a chunk.  For resuming, in all protocols I have seen this can only be done by chunk (you re-upload the chunk that was interrupted, you don't actually resume the chunk) so an accurate 'bytes sent' is not required.

**Comment by escamoteur** [comment]
But that's only if I can send it in chunks without caring of the order of the chunks. Have to see if mux allows that. I was hoping to use the Task Progress to know which part definitely has been uploaded to start  the next try from where the previous failed. That's why I would need absolute bytes and no percentage
Am 25. Okt. 2024, 20:44 +0100 schrieb 781flyingdutchman ***@***.***>:

**Comment by 781flyingdutchman** [comment]
Got it - I think that's why the AWS and GCP protocols do use numbered chunks. Either way, I think `progress * expectedFileSize` should work, as a `double` is plenty accurate when multiplied with a chunk up to several MB in length I am sure.

**Comment by escamoteur** [comment]
Sorry to disagree, if you are just one byte off you can get into real problems depending on the file type.
And you have the absolute bytes internally otherwise you couldn't calculate the percentage. So it would be great if that could be added to the progress status.
Am 25. Okt. 2024, 21:01 +0100 schrieb 781flyingdutchman ***@***.***>:

**Comment by 781flyingdutchman** [comment]
Have to disagree wit that: the information is there - you are assuming a lack of precision in multiplying a double with an int that I am quite sure is not warranted. Just try it, and you'll see the calculated number of bytes sent is within a hundredth of a byte, so can easily be rounded.

**Comment by 781flyingdutchman** [comment]
Also, if you're trying to use # bytes sent I think you are taking a dangerous approach: if an error occurs, then I am not sure exactly where that error occurred - it may be after the past progress update (so the server received more bytes than you think) and depending on OS implementation it may even be that if the connection is interrupted the # bytes sent into the send buffer is greater than what actually made it to the server. IMHO you should never rely on the client side to tell you whas has been received on the server side. That is also why typical chunk upload protocols work on a chunk by chunk basis (no partial chunks - just resent the whole chunk if something went wrong) and wait for acknowledgement of the full chunk before marking it complete.

**Comment by escamoteur** [comment]
I admit that's a very valid point. I currently store the chunk number before the next chunk ist started.
I thought that when a progress update is made that means that this part was received by the server not just put into the client FIFO.

If I queue more than one chunk, are they sent sequentially and in order or in parallel?
Am 25. Okt. 2024, 22:30 +0100 schrieb 781flyingdutchman ***@***.***>:

**Comment by 781flyingdutchman** [comment]
I think it's a grey area, depending on how the OS exactly implements its HTTP client. That's why you use the HTTP response (one level up from packet transport) for the server to acknowledge receipt and processing of the request. That's the level you should build your retries etc around.  Similarly on downloads, it's the client that determines where the resume should restart - not the server (sender).

Multiple queues can be processed in parallel, in any order, though you can conceivably use queues and `maxConcurretnByHost = 1` to force sequencing, but again this is not robust to rely on, which is why AWS and GCP use numbered chunks.

**Comment by escamoteur** [comment]
Yeah, I have to check if mux allows out of order chunks.
Am 25. Okt. 2024, 22:45 +0100 schrieb 781flyingdutchman ***@***.***>:
